packages/by-name/kata/kata-runtime/0020-genpolicy-support-dynamic-annotations.patch

```diff
@@ -0,0 +1,149 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Moritz Sanft <58110325+msanft@users.noreply.github.com>
+Date: Mon, 20 Jan 2025 13:44:00 +0100
+Subject: [PATCH] genpolicy: support dynamic annotations
+
+This adds support for handling annotations with dynamic keys to
+genpolicy. This is necessary for use-cases like GPU containers, where
+in-cluster components (i.e. post policy-generation) instrument
+containers with annotations with varying keys, like `cdi.k8s.io/vfioXY`,
+where `XY` corresponds to a dynamic ID.
+---
+ src/tools/genpolicy/genpolicy-settings.json |  8 ++-
+ src/tools/genpolicy/rules.rego              | 56 +++++++++++++++++++--
+ src/tools/genpolicy/src/policy.rs           |  6 +++
+ 3 files changed, 64 insertions(+), 6 deletions(-)
+
+diff --git a/src/tools/genpolicy/genpolicy-settings.json b/src/tools/genpolicy/genpolicy-settings.json
+index 9b95f9f7462717d04f0b9ce685d97c0455f949da..7dac0e5e0585c25e324a39656d1a2dcfa12e7d96 100644
+--- a/src/tools/genpolicy/genpolicy-settings.json
++++ b/src/tools/genpolicy/genpolicy-settings.json
+@@ -309,7 +309,10 @@
+             "CAP_PERFMON",
+             "CAP_BPF",
+             "CAP_CHECKPOINT_RESTORE"
+-        ]
++        ],
++        "dynamic_annotations": {
++            "^cdi\\.k8s\\.io\\/vfio[0-9]{2}$": "^nvidia.com/gpu=[0-9]+$"
++        }
+     },
+     "kata_config": {
+         "confidential_guest": false,
+@@ -333,7 +336,8 @@
+                 "^AZURE_CLIENT_ID=[A-Fa-f0-9-]*$",
+                 "^AZURE_TENANT_ID=[A-Fa-f0-9-]*$",
+                 "^AZURE_FEDERATED_TOKEN_FILE=/var/run/secrets/azure/tokens/azure-identity-token$",
+-                "^AZURE_AUTHORITY_HOST=https://login\\.microsoftonline\\.com/$"
++                "^AZURE_AUTHORITY_HOST=https://login\\.microsoftonline\\.com/$",
++                "^PCI_RESOURCE_NVIDIA_COM.*=[a-fA-F0-9:.-]*$"
+             ]
+         },
+         "CopyFileRequest": [
+diff --git a/src/tools/genpolicy/rules.rego b/src/tools/genpolicy/rules.rego
+index 43cb19a56fe8ea5833708f0639c9e85ddd884cb3..271df2aebe05bd4bbd7aa396be24fb6fee0668bf 100644
+--- a/src/tools/genpolicy/rules.rego
++++ b/src/tools/genpolicy/rules.rego
+@@ -199,26 +199,31 @@ allow_anno(p_oci, i_oci) {
+ }
+ allow_anno(p_oci, i_oci) {
+     print("allow_anno 2: p Annotations =", p_oci.Annotations)
++    p_dynamic_annotations := policy_data.common.dynamic_annotations
```

Comment:
Instead of plumbing this through all of the predicates, I'd just access the global where you actually need it.